### DISCUSSION TITLE
Provide host simulation start time as an option in shadow config file

### DISCUSSION THREAD

**Discussion by jerry855202** [shadow/shadow#3328] [Ideas]
Currently, the start time within each node is hard-coded by ‎EMUTIME_SIMULATION_START in emulated_time.rs to be 00:00:00 UTC, 1 January, 2000.
Right now if one have to change the start time within the node environment, they'd have to change this constant and recompile each time. It seems to me that it might be useful to add this as an environment option.
For example, this would probably be useful for people testing w/ things like SSL certificates.

**Comment by robgjansen** [comment]
Interesting idea, thanks for raising!

@stevenengler @sporksmith can you think of anything that would break *inside of Shadow* if we made `EMUTIME_SIMULATION_START` configurable and the user chose a weird time? Of course they might choose a time that breaks their own apps they are trying to run in Shadow, but I think that's OK as long as Shadow itself can gracefully handle different times.

**Reply by sporksmith** [reply]
Not specifically, but TBH I'd be a little surprised if nothing cropped up heh

I *have* worked around this in the past by giving the applications multi-year `start_time`s; I think combined with options to minimize logging and heartbeats, but I agree that a configurable simulation start time would be much nicer.

**Reply by stevenengler** [reply]
Without looking at the code, some initial thoughts are:

1. This is currently a constant. Do we currently use this in any const contexts (for example const functions)? If so, is this difficult to change?
2. We have functions like `EmulatedTime::from_abs_simtime(x: SimulationTime)` to convert between the two time types. But with a user-configurable start time, either that function would need another argument for the start time from the configuration file which might require shadow to add a bunch of plumbing to pass that value around, or we'd need to put this start time option in a `OnceCell` (or maybe just atomic) global at the start of the simulation and access it from within `from_abs_simtime`.
3. As long as the user doesn't choose anything near `u64::MAX` it probably would be fine. In this case the C code might have issues but the rust code should panic if it overflows. But if there are any syscall handlers that need to use 32-bit time types (I'm not sure if there are), we wouldn't guarantee to handle those sanely or how Linux does if they set the year to >=2038. Shadow would probably panic in this case.
4. How should we expose this in the configuration file? I don't think we want to allow the user to enter an ISO formatted time or anything, since that would require a date/time parsing library and we don't want to have to worry about timezones. I think we can just let the user enter an integer number of seconds since the Unix epoch?

**Reply by robgjansen** [reply]
For 3., could we validate the user input and consider the configuration invalid if the simulation end time overflows (is >=2038)?

For 4., an integer number of seconds since the Unix epoch seems fine - it's simple and would hopefully lead to fewer bugs.

**Reply by sporksmith** [reply]
I think we'll also need to plumb this through to the shim side, particularly for the shim logger. It'd probably make sense to add it to the Manager's shared memory struct.

**Reply by sporksmith** [reply]
> For 3., could we validate the user input and consider the configuration invalid if the simulation end time overflows (is >=2038)?

I think this is somewhat orthogonal to the parent feature request; e.g. a user could already set a multi-year end time to end up in > 2038. FWIW I'd also lean towards continuing to allow it and to start shaking out those bugs since 2038 is only 14 years in the future (!!), and shadow could be useful for testing other software for 2038 bugs.

**Reply by stevenengler** [reply]
From looking at the code, it seems like a fair amount of work to plumb this option around everywhere (`EMUTIME_SIMULATION_START` is also aliased as `EmulatedTime::SIMULATION_START`). We also have some other constants that indirectly depend on this value:

```rust
// cbindgen refuses to do the arithmetic here, so we we pre-compute,
// and validate in the static assertion below.
pub const SIMTIME_MAX: CSimulationTime = 17500059273709551614u64;
const _: () =
    assert!(SIMTIME_MAX == emulated_time::EMUTIME_MAX - emulated_time::EMUTIME_SIMULATION_START);
```

If this option isn't needed often, I think it would be a much easier path to just recompile as needed, or to generate the TLS certs using something like [faketime](https://manpages.ubuntu.com/manpages/trusty/man1/faketime.1.html). Otherwise yeah I think making this a configuration option is possible, but it would be a bit painful and would take some work.

**Reply by jerry855202** [reply]
I'm not particularly familiar with the code base yet, so please bear with me if I asked something trivial.

If it would be easier to recompile on-deamnd, are we able to get the desired result just by changing the value for ‎SIMULATION_START_SEC and EMUTIME_SIMULATION_START‎ in emulated_time.rs, on line 27 and 36 respectrively?

I've searched the codebase very briefly for the two symbols, and also 946684800u64, and from what I can tell, everything else is referenced from these values.

**Reply by stevenengler** [reply]
> If it would be easier to recompile on-deamnd, are we able to get the desired result just by changing the value for ‎SIMULATION_START_SEC and EMUTIME_SIMULATION_START‎ in emulated_time.rs, on line 27 and 36 respectrively?

I think you'll need to update those two values and the value of `SIMTIME_MAX` in `src/lib/shadow-shim-helper-rs/src/simulation_time.rs`, otherwise an [assertion](https://github.com/shadow/shadow/blob/62b6b2be7e054a033faad6c947f0237b6150ccfc/src/lib/shadow-shim-helper-rs/src/simulation_time.rs#L488-L492) will fail. Otherwise I think that's all you need to change. Make sure to run `./setup test` after to make sure nothing broke.

> I'm not particularly familiar with the code base yet, so please bear with me if I asked something trivial.

It's not trivial so don't worry. The reason it's overly complicated is because we need to expose these constants in both Shadow's C code and Rust code (and to do so in a way that works with cbindgen). As Shadow moves more code to Rust, hopefully we can simplify this some day.

**Comment by jerry855202** [comment]
> I think you'll need to update those two values and the value of `SIMTIME_MAX` in `src/lib/shadow-shim-helper-rs/src/simulation_time.rs`, otherwise an [assertion](https://github.com/shadow/shadow/blob/62b6b2be7e054a033faad6c947f0237b6150ccfc/src/lib/shadow-shim-helper-rs/src/simulation_time.rs#L488-L492) will fail. Otherwise I think that's all you need to change. Make sure to run `./setup test` after to make sure nothing broke.

Thank you. I'll try to patch these values out, and if anything interesting comes up, I'll update them in this thread.
